File: backend/app/services/tee_times/mock.py
"""
MockTeeTimeProvider — Phase 1 placeholder.

Generates deterministic, realistic availability for a small set of courses.
Clearly labelled provider="mock" so the API response can be labelled "Demo data."

Cache-first: results are cached per query key for the lifetime of the process.
A real provider replaces this with an HTTP call + cache layer (e.g. Redis / in-memory
TTL cache) — the cache key and slot shape stay the same.

This is the seam real providers slot into:
  - Replace MockTeeTimeProvider in tee_times.py with ChronogolfProvider (or GolfNowProvider).
  - The route, schema, and UI require no changes.
"""

from __future__ import annotations

import json
from urllib.parse import quote

from .base import (
    BookingDetails,
    BookingResult,
    TeeTimeProvider,
    TeeTimeQuery,
    TeeTimeSlot,
)


# ─── Course catalogue ──────────────────────────────────────────────────────────

_COURSES = [
    dict(id="presidio",       name="Presidio Golf Course",            city="San Francisco, CA", dist=4.1,  rating=4.3, designer="Robert Trent Jones Jr.", base_price=86,  cart=False, holes=18),
    dict(id="harding",        name="Harding Park",                    city="San Francisco, CA", dist=6.8,  rating=4.5, designer="Willie Watson",           base_price=145, cart=False, holes=18),
    dict(id="lincoln",        name="Lincoln Park Golf Course",         city="San Francisco, CA", dist=5.2,  rating=3.9, designer=None,                      base_price=52,  cart=False, holes=18),
    dict(id="sharp",          name="Sharp Park Golf Course",           city="Pacifica, CA",      dist=12.4, rating=3.6, designer="Alister MacKenzie",       base_price=38,  cart=False, holes=18),
    dict(id="bethpage-black", name="Bethpage State Park — Black",     city="Farmingdale, NY",   dist=31.2, rating=4.8, designer="A.W. Tillinghast",         base_price=95,  cart=False, holes=18),
    dict(id="crystal-springs",name="Crystal Springs Golf Course",     city="San Bruno, CA",     dist=16.7, rating=3.8, designer=None,                      base_price=65,  cart=True,  holes=18),
]


# ─── Helpers ───────────────────────────────────────────────────────────────────

def _to_minutes(hhmm: str) -> int:
    h, m = (int(x) for x in hhmm.split(":"))
    return h * 60 + m


def _from_minutes(mins: int) -> str:
    h = (mins // 60) % 24
    m = mins % 60
    return f"{h:02d}:{m:02d}"


def _seeded_random(seed_str: str):
    """LCG PRNG seeded from a string — deterministic, not cryptographic."""
    h = 0
    for ch in seed_str:
        h = (31 * h + ord(ch)) & 0xFFFF_FFFF
    state = [h]

    def _rand() -> float:
        state[0] = (1_664_525 * state[0] + 1_013_904_223) & 0xFFFF_FFFF
        return state[0] / 0x1_0000_0000

    return _rand


def _cache_key(query: TeeTimeQuery) -> str:
    return json.dumps({
        "date": query.date,
        "start": query.time_window_start,
        "end": query.time_window_end,
        "party": query.party_size,
        "ids": sorted(query.course_ids),
        "area": query.area,
    }, sort_keys=True)


# ─── Provider ──────────────────────────────────────────────────────────────────

_CACHE: dict[str, list[TeeTimeSlot]] = {}


class MockTeeTimeProvider(TeeTimeProvider):
    """Phase 1 mock — generates deterministic availability; never calls the network."""

    @property
    def name(self) -> str:
        return "mock"

    async def search_availability(self, query: TeeTimeQuery) -> list[TeeTimeSlot]:
        key = _cache_key(query)
        if key in _CACHE:
            return _CACHE[key]
        slots = _generate(query)
        _CACHE[key] = slots
        return slots

    async def book(self, slot: TeeTimeSlot, _details: BookingDetails) -> BookingResult:
        suffix = slot.id[-6:].upper()
        return BookingResult(
            status="confirmed",
            confirmation_number=f"MOCK-{suffix}",
            message="Mock booking confirmed. No real reservation was made.",
        )


def _generate(query: TeeTimeQuery) -> list[TeeTimeSlot]:
    rng = _seeded_random(_cache_key(query))
    start_min = _to_minutes(query.time_window_start)
    end_min = _to_minutes(query.time_window_end)
    window = end_min - start_min
    if window <= 0:
        return []

    eligible = [
        c for c in _COURSES
        if (not query.course_ids or c["id"] in query.course_ids)
        and (query.max_distance_miles is None or c["dist"] <= query.max_distance_miles)
    ]

    slots: list[TeeTimeSlot] = []
    for course in eligible:
        count = 2 + int(rng() * 3)
        used: set[int] = set()
        for i in range(count):
            # 8-minute interval slots.
            intervals = max(1, window // 8)
            offset = int(rng() * intervals) * 8
            while offset in used and len(used) < intervals:
                offset = (offset + 8) % (intervals * 8)
            used.add(offset)
            tee_min = start_min + offset
            if tee_min >= end_min:
                continue

            avail = int(rng() * (4 - query.party_size + 1))
            player_slots = min(4, max(query.party_size, query.party_size + avail))

            price_var = 0.85 + rng() * 0.30
            price_usd = round(course["base_price"] * price_var, 2)
            if query.max_price_usd is not None and price_usd > query.max_price_usd:
                continue

            time_str = _from_minutes(tee_min)
            slot_id = f"{course['id']}-{query.date}-{time_str}-{i}"

            slots.append(TeeTimeSlot(
                id=slot_id,
                course_id=course["id"],
                course_name=course["name"],
                city=course["city"],
                date=query.date,
                time=time_str,
                players=player_slots,
                price_usd=price_usd,
                cart_included=course["cart"],
                distance_miles=course["dist"],
                rating=course["rating"],
                designer=course.get("designer"),
                provider="mock",
                holes=course["holes"],  # type: ignore[arg-type]
                booking_url=f"https://www.golfnow.com/tee-times/facility/{quote(course['name'])}/search",
            ))

    # Sort by distance, then time.
    slots.sort(key=lambda s: (s.distance_miles, s.time))
    return slots

File: frontend/src/lib/teetime/providers/mock.ts
/**
 * MockTeeTimeProvider — Phase 1 placeholder.
 *
 * Generates deterministic, realistic availability for a small set of courses.
 * Clearly labelled `provider: "mock"` so the UI can show "Demo data" when
 * this provider is active.
 *
 * Seam for real providers: replace this file's usage in the registry with a
 * real provider (e.g. `ChronogolfProvider`) that calls the Lightspeed Partner
 * API, and the UI + backend surface work unchanged.
 *
 * Cache-first: once a query key has been resolved the result is stored in a
 * module-level Map and returned immediately on repeat calls (matching the
 * GolfAPI cache-first pattern in golf-api.ts).
 */

import type { TeeTimeProvider } from "../provider";
import type { TeeTimeQuery, TeeTimeSlot, BookingDetails, BookingResult } from "../types";

// ─── Course catalogue (mock) ──────────────────────────────────────────────────

interface MockCourse {
  id: string;
  name: string;
  city: string;
  distanceMiles: number;
  rating: number;
  designer?: string;
  basePrice: number;
  cartIncluded: boolean;
  holes: 9 | 18;
}

const MOCK_COURSES: MockCourse[] = [
  {
    id: "presidio",
    name: "Presidio Golf Course",
    city: "San Francisco, CA",
    distanceMiles: 4.1,
    rating: 4.3,
    designer: "Robert Trent Jones Jr.",
    basePrice: 86,
    cartIncluded: false,
    holes: 18,
  },
  {
    id: "harding",
    name: "Harding Park",
    city: "San Francisco, CA",
    distanceMiles: 6.8,
    rating: 4.5,
    designer: "Willie Watson",
    basePrice: 145,
    cartIncluded: false,
    holes: 18,
  },
  {
    id: "lincoln",
    name: "Lincoln Park Golf Course",
    city: "San Francisco, CA",
    distanceMiles: 5.2,
    rating: 3.9,
    basePrice: 52,
    cartIncluded: false,
    holes: 18,
  },
  {
    id: "sharp",
    name: "Sharp Park Golf Course",
    city: "Pacifica, CA",
    distanceMiles: 12.4,
    rating: 3.6,
    designer: "Alister MacKenzie",
    basePrice: 38,
    cartIncluded: false,
    holes: 18,
  },
  {
    id: "bethpage-black",
    name: "Bethpage State Park — Black",
    city: "Farmingdale, NY",
    distanceMiles: 31.2,
    rating: 4.8,
    designer: "A.W. Tillinghast",
    basePrice: 95,
    cartIncluded: false,
    holes: 18,
  },
  {
    id: "crystal-springs",
    name: "Crystal Springs Golf Course",
    city: "San Bruno, CA",
    distanceMiles: 16.7,
    rating: 3.8,
    basePrice: 65,
    cartIncluded: true,
    holes: 18,
  },
];

// ─── Helpers ──────────────────────────────────────────────────────────────────

/** Parse "HH:MM" into total minutes since midnight. */
function toMinutes(hhmm: string): number {
  const [h, m] = hhmm.split(":").map(Number);
  return h * 60 + (m || 0);
}

/** Format total minutes since midnight to "HH:MM". */
function fromMinutes(mins: number): string {
  const h = Math.floor(mins / 60) % 24;
  const m = mins % 60;
  return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}`;
}

/**
 * Generate a stable pseudo-random number in [0, 1) from a string seed.
 * Deterministic so the same query always returns the same "availability."
 */
function seededRandom(seed: string): () => number {
  // Simple hash → 32-bit integer.
  let h = 0;
  for (let i = 0; i < seed.length; i++) {
    h = (Math.imul(31, h) + seed.charCodeAt(i)) | 0;
  }
  // LCG PRNG
  let state = h;
  return () => {
    state = (Math.imul(1664525, state) + 1013904223) | 0;
    return ((state >>> 0) / 0x100000000);
  };
}

// ─── Cache ────────────────────────────────────────────────────────────────────

const _cache = new Map<string, TeeTimeSlot[]>();

function cacheKey(query: TeeTimeQuery): string {
  return JSON.stringify({
    date: query.date,
    start: query.timeWindowStart,
    end: query.timeWindowEnd,
    partySize: query.partySize,
    courseIds: query.courseIds?.slice().sort(),
    area: query.area,
    maxDistanceMiles: query.maxDistanceMiles ?? null,
    maxPriceUsd: query.maxPriceUsd ?? null,
  });
}

// ─── Mock Provider ────────────────────────────────────────────────────────────

export class MockTeeTimeProvider implements TeeTimeProvider {
  readonly name = "mock";

  async searchAvailability(query: TeeTimeQuery): Promise<TeeTimeSlot[]> {
    const key = cacheKey(query);
    if (_cache.has(key)) return _cache.get(key)!;

    const slots = this._generate(query);
    _cache.set(key, slots);
    return slots;
  }

  async book(slot: TeeTimeSlot, _details: BookingDetails): Promise<BookingResult> {
    // Mock booking: always succeeds.  Real providers replace this with an API call.
    const suffix = slot.id.slice(-6).toUpperCase();
    return {
      status: "confirmed",
      confirmationNumber: `MOCK-${suffix}`,
      message: "Mock booking confirmed. No real reservation was made.",
    };
  }

  // ─── Internal generation ───────────────────────────────────────────────────

  private _generate(query: TeeTimeQuery): TeeTimeSlot[] {
    const rng = seededRandom(cacheKey(query));
    const startMin = toMinutes(query.timeWindowStart);
    const endMin = toMinutes(query.timeWindowEnd);

    // Filter courses by courseIds / maxDistanceMiles.
    const eligible = MOCK_COURSES.filter((c) => {
      if (query.courseIds?.length && !query.courseIds.includes(c.id)) return false;
      if (query.maxDistanceMiles != null && c.distanceMiles > query.maxDistanceMiles) return false;
      return true;
    });

    const slots: TeeTimeSlot[] = [];

    for (const course of eligible) {
      // Generate 2–4 tee times spread across the window.
      const count = 2 + Math.floor(rng() * 3); // 2, 3, or 4
      const windowSpan = endMin - startMin;
      const used = new Set<number>();

      for (let i = 0; i < count; i++) {
        // Pick a time in the window, rounded to 8-minute intervals.
        let offset = Math.floor(rng() * (windowSpan / 8)) * 8;
        // Avoid exact duplicates.
        while (used.has(offset)) offset = (offset + 8) % windowSpan;
        used.add(offset);
        const teeMin = startMin + offset;
        if (teeMin >= endMin) continue;

        // Available slots: must be >= partySize.
        const available = query.partySize + Math.floor(rng() * (4 - query.partySize + 1));
        const playerSlots = Math.min(4, Math.max(query.partySize, available));

        // Price variation ±15%.
        const priceVariation = 0.85 + rng() * 0.30;
        const priceUsd = Math.round(course.basePrice * priceVariation);

        // Skip if above ceiling.
        if (query.maxPriceUsd != null && priceUsd > query.maxPriceUsd) continue;

        const time = fromMinutes(teeMin);
        const slotId = `${course.id}-${query.date}-${time}-${i}`;

        slots.push({
          id: slotId,
          courseId: course.id,
          courseName: course.name,
          city: course.city,
          date: query.date,
          time,
          players: playerSlots,
          priceUsd,
          cartIncluded: course.cartIncluded,
          distanceMiles: course.distanceMiles,
          rating: course.rating,
          designer: course.designer,
          provider: "mock",
          holes: course.holes,
          // Phase 1: deep-link to GolfNow search for this course.
          bookingUrl: `https://www.golfnow.com/tee-times/facility/${encodeURIComponent(course.name)}/search`,
        });
      }
    }

    // Sort by distance, then time.
    slots.sort((a, b) => a.distanceMiles - b.distanceMiles || a.time.localeCompare(b.time));
    return slots;
  }
}

/** Singleton instance — import this where you need the mock. */
export const mockProvider = new MockTeeTimeProvider();
